NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -1,45 +1,143 @@
+# This file defines the abstract type AbstractCategory
+# all fusion categories (Z{2}, SU2, Ising...) are subtypes of AbstractCategory
+
 abstract type AbstractCategory end
 
-label(c::AbstractCategory) = error("method `label` not defined for type $(typeof(c))")
+# ============  Base interface  =================
+function Base.isless(c1::C, c2::C) where {C<:AbstractCategory}
+  return isless(category_label(c1), category_label(c2))
+end
 
-function dimension(c::AbstractCategory)
-  return error("method `dimension` not defined for type $(typeof(c))")
+# =================  Misc  ======================
+trivial(x) = trivial(typeof(x))
+function trivial(axis_type::Type{<:AbstractUnitRange})
+  return GradedAxes.gradedrange([trivial(eltype(axis_type))])  # always returns nondual
+end
+function trivial(la_type::Type{<:LabelledNumbers.LabelledInteger})
+  return la_type(1, trivial(LabelledNumbers.label_type(la_type)))
+end
+function trivial(type::Type)
+  return error("`trivial` not defined for type $(type).")
 end
 
-function label_fusion_rule(category_type::Type{<:AbstractCategory}, l1, l2)
-  return error("`label_fusion_rule` not defined for type $(category_type).")
+istrivial(c::AbstractCategory) = (c == trivial(typeof(c)))
+
+function category_label(c::AbstractCategory)
+  return error("method `category_label` not defined for type $(typeof(c))")
+end
+
+function GradedAxes.dual(category_type::Type{<:AbstractCategory})
+  return error("`dual` not defined for type $(category_type).")
+end
+
+block_boundaries(g::AbstractUnitRange) = block_boundaries(SymmetryStyle(g), g)
+block_boundaries(::AbelianGroup, g) = GradedAxes.unlabel.(BlockArrays.blocklengths(g))
+function block_boundaries(::NonAbelianGroup, g)
+  return Sectors.quantum_dimension.(GradedAxes.blocklabels(g)) .*
+         BlockArrays.blocklengths(g)
+end
+
+quantum_dimension(x) = quantum_dimension(SymmetryStyle(x), x)
+
+function quantum_dimension(::SymmetryStyle, c::AbstractCategory)
+  return error("method `quantum_dimension` not defined for type $(typeof(c))")
 end
 
-function fusion_rule(c1::AbstractCategory, c2::AbstractCategory)
-  category_type = typeof(c1)
-  return [category_type(l) for l in label_fusion_rule(category_type, label(c1), label(c2))]
+quantum_dimension(::AbelianGroup, ::AbstractCategory) = 1
+quantum_dimension(::EmptyCategory, ::AbstractCategory) = 0
+
+function quantum_dimension(::SymmetryStyle, g::AbstractUnitRange)
+  gblocks = BlockArrays.blocklengths(g)
+  return sum(gblocks .* quantum_dimension.(LabelledNumbers.label.(gblocks)))
```

Comment:
Is this function just `sum(block_boundaries(g))`? If so that would be a simpler definition and generic to all symmetry styles.

Reply:
Historical reasons, this function is older than `block_dimensions`. I just need to extend `block_dimensions` to `NonGroupCategory` - I did not need it in that case but it still makes sense.